Security/Guidelines/OpenID_connect.mediawiki

```diff
@@ -0,0 +1,177 @@
+| The Refresh token enables it's bearer to request and obtain new Access tokens. These newly obtained access tokens have a subset of the permissions that the Refresh token has. The Refresh token never expires.
+|-
+! '''ID token'''
+| Base64-encoded JSON document ([https://jwt.io/ JWT])
+| The ID token contains information about how and when the user authenticated along with various attributes. ID tokens are created and signed by OpenID Connect Providers (OP) and consumed, verified by web applications authenticating users (RPs).
+|-
+|}
+
+==== Detailed OIDC authentication flow ====
+
+This sequence diagram is useful if you want to understand how OIDC works, or need to modify or implement your own OIDC library.
+
+[[File:Authentication_Sequence_with_OpenID_Connect(1).png|1100px]]
+
+== Implement authentication with OpenID Connect (OIDC) securely in my web applications (RP) ==
+
+=== Session handling ===
+
+The OpenID Connect Provider (OP) typically creates a user session cookie so that it does not need to re-ask the user for their
+credentials too often across different web applications (RP).
+The expiration of the session depends on the OP setup the session and may be forced to expire by the OpenID Connect Provider (OP) sooner than the cookie indicates on the user's browser.
+This allows the OP to forcibly log the user out from the OP point of view. This premature termination of the user’s session with the OP will not, however, end the user’s session on the web application's (RP's) which they’ve logged into.
+
+For that reason, it is important that the web application (RP) respect the following set of rules in regards to session handling:
+
+
+* '''Must''' expire the user session when the ID token expires or sooner (the expiration time is generally a UNIX timestamp attribute named <code>exp</code>).
+* If the user's session is longer than '''15 minutes''', '''must''' re-check/introspect the ID token every ''15 minutes'' or next user request (whichever comes first), to ensure that the user is still valid and has correct permissions.
+** This ensures that access is revoked within ''15 minutes'' in the event that the user's account is disabled by the OpenID Connect Provider (OP).
+** This issues a new ID token, which new attributes if they have changed.
+** This usually also renews the ID token, which will update it's expiration time to a later date.
+* '''Optionally''' provide a <code>logout</code> URL, which the OpenID Connect Provider (OP) can call to indicate if a user has logged out (so that the web application immediately know when to log the user out as well).
+
+=== Other important security considerations ===
+
+'''ID Token''':
+
+* '''Always''' verify the id token signature.
+* '''Always''' expire the user session when the associated ID token expires.
+* '''Should''' introspect the contents of the id token by querying the OP regularly, before the ID token expires.
+
+'''Authorization Code Grant''':
```

Comment:
I mentioned this in the previous iteration of the doc, but this is the first time the reader comes upon this phrase. You should say what it is or means since otherwise it makes no sense.